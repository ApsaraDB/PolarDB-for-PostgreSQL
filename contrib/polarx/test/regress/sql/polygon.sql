--
-- POLYGON
--
-- polygon logic
--

CREATE TABLE POLYGON_TBL(ID int, f1 polygon);


INSERT INTO POLYGON_TBL(ID,f1) VALUES (1, '(2.0,0.0),(2.0,4.0),(0.0,0.0)');

INSERT INTO POLYGON_TBL(ID,f1) VALUES (2, '(3.0,1.0),(3.0,3.0),(1.0,0.0)');

-- degenerate polygons 
INSERT INTO POLYGON_TBL(ID,f1) VALUES (3, '(0.0,0.0)');

INSERT INTO POLYGON_TBL(ID,f1) VALUES (4, '(0.0,1.0),(0.0,1.0)');

-- bad polygon input strings 
INSERT INTO POLYGON_TBL(ID,f1) VALUES (5, '0.0');

INSERT INTO POLYGON_TBL(ID,f1) VALUES (6, '(0.0 0.0');

INSERT INTO POLYGON_TBL(ID,f1) VALUES (7, '(0,1,2)');

INSERT INTO POLYGON_TBL(ID,f1) VALUES (8, '(0,1,2,3');

INSERT INTO POLYGON_TBL(ID,f1) VALUES (9, 'asdf');


SELECT '' AS four, * FROM POLYGON_TBL ORDER BY ID;

-- overlap
SELECT '' AS three, p.*
   FROM POLYGON_TBL p
   WHERE p.f1 && '(3.0,1.0),(3.0,3.0),(1.0,0.0)' ORDER BY ID;

-- left overlap
SELECT '' AS four, p.*
   FROM POLYGON_TBL p
   WHERE p.f1 &< '(3.0,1.0),(3.0,3.0),(1.0,0.0)' ORDER BY ID;

-- right overlap
SELECT '' AS two, p.*
   FROM POLYGON_TBL p
   WHERE p.f1 &> '(3.0,1.0),(3.0,3.0),(1.0,0.0)' ORDER BY ID;

-- left of
SELECT '' AS one, p.*
   FROM POLYGON_TBL p
   WHERE p.f1 << '(3.0,1.0),(3.0,3.0),(1.0,0.0)' ORDER BY ID;

-- right of
SELECT '' AS zero, p.*
   FROM POLYGON_TBL p
   WHERE p.f1 >> '(3.0,1.0),(3.0,3.0),(1.0,0.0)' ORDER BY ID;

-- contained
SELECT '' AS one, p.*
   FROM POLYGON_TBL p
   WHERE p.f1 <@ polygon '(3.0,1.0),(3.0,3.0),(1.0,0.0)' ORDER BY ID;

-- same
SELECT '' AS one, p.*
   FROM POLYGON_TBL p
   WHERE p.f1 ~= polygon '(3.0,1.0),(3.0,3.0),(1.0,0.0)' ORDER BY ID;

-- contains
SELECT '' AS one, p.*
   FROM POLYGON_TBL p
   WHERE p.f1 @> polygon '(3.0,1.0),(3.0,3.0),(1.0,0.0)' ORDER BY ID;

--
-- polygon logic
--
-- left of
SELECT polygon '(2.0,0.0),(2.0,4.0),(0.0,0.0)' << polygon '(3.0,1.0),(3.0,3.0),(1.0,0.0)' AS false;

-- left overlap
SELECT polygon '(2.0,0.0),(2.0,4.0),(0.0,0.0)' << polygon '(3.0,1.0),(3.0,3.0),(1.0,0.0)' AS true;

-- right overlap
SELECT polygon '(2.0,0.0),(2.0,4.0),(0.0,0.0)' &> polygon '(3.0,1.0),(3.0,3.0),(1.0,0.0)' AS false;

-- right of
SELECT polygon '(2.0,0.0),(2.0,4.0),(0.0,0.0)' >> polygon '(3.0,1.0),(3.0,3.0),(1.0,0.0)' AS false;

-- contained in
SELECT polygon '(2.0,0.0),(2.0,4.0),(0.0,0.0)' <@ polygon '(3.0,1.0),(3.0,3.0),(1.0,0.0)' AS false;

-- contains
SELECT polygon '(2.0,0.0),(2.0,4.0),(0.0,0.0)' @> polygon '(3.0,1.0),(3.0,3.0),(1.0,0.0)' AS false;

SELECT '((0,4),(6,4),(1,2),(6,0),(0,0))'::polygon @> '((2,1),(2,3),(3,3),(3,1))'::polygon AS "false";

SELECT '((0,4),(6,4),(3,2),(6,0),(0,0))'::polygon @> '((2,1),(2,3),(3,3),(3,1))'::polygon AS "true";

SELECT '((1,1),(1,4),(5,4),(5,3),(2,3),(2,2),(5,2),(5,1))'::polygon @> '((3,2),(3,3),(4,3),(4,2))'::polygon AS "false";

SELECT '((0,0),(0,3),(3,3),(3,0))'::polygon @> '((2,1),(2,2),(3,2),(3,1))'::polygon AS "true";

-- same
SELECT polygon '(2.0,0.0),(2.0,4.0),(0.0,0.0)' ~= polygon '(3.0,1.0),(3.0,3.0),(1.0,0.0)' AS false;

-- overlap
SELECT polygon '(2.0,0.0),(2.0,4.0),(0.0,0.0)' && polygon '(3.0,1.0),(3.0,3.0),(1.0,0.0)' AS true;

SELECT '((0,4),(6,4),(1,2),(6,0),(0,0))'::polygon && '((2,1),(2,3),(3,3),(3,1))'::polygon AS "true";

SELECT '((1,4),(1,1),(4,1),(4,2),(2,2),(2,4),(1,4))'::polygon && '((3,3),(4,3),(4,4),(3,4),(3,3))'::polygon AS "false";
SELECT '((200,800),(800,800),(800,200),(200,200))' &&  '(1000,1000,0,0)'::polygon AS "true";

-- distance from a point
SELECT	'(0,0)'::point <-> '((0,0),(1,2),(2,1))'::polygon as on_corner,
	'(1,1)'::point <-> '((0,0),(2,2),(1,3))'::polygon as on_segment,
	'(2,2)'::point <-> '((0,0),(1,4),(3,1))'::polygon as inside,
	'(3,3)'::point <-> '((0,2),(2,0),(2,2))'::polygon as near_corner,
	'(4,4)'::point <-> '((0,0),(0,3),(4,0))'::polygon as near_segment;
