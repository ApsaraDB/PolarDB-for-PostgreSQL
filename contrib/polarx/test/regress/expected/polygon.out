--
-- POLYGON
--
-- polygon logic
--
CREATE TABLE POLYGON_TBL(ID int, f1 polygon);
INSERT INTO POLYGON_TBL(ID,f1) VALUES (1, '(2.0,0.0),(2.0,4.0),(0.0,0.0)');
INSERT INTO POLYGON_TBL(ID,f1) VALUES (2, '(3.0,1.0),(3.0,3.0),(1.0,0.0)');
-- degenerate polygons 
INSERT INTO POLYGON_TBL(ID,f1) VALUES (3, '(0.0,0.0)');
INSERT INTO POLYGON_TBL(ID,f1) VALUES (4, '(0.0,1.0),(0.0,1.0)');
-- bad polygon input strings 
INSERT INTO POLYGON_TBL(ID,f1) VALUES (5, '0.0');
ERROR:  invalid input syntax for type polygon: "0.0"
LINE 1: INSERT INTO POLYGON_TBL(ID,f1) VALUES (5, '0.0');
                                                  ^
INSERT INTO POLYGON_TBL(ID,f1) VALUES (6, '(0.0 0.0');
ERROR:  invalid input syntax for type polygon: "(0.0 0.0"
LINE 1: INSERT INTO POLYGON_TBL(ID,f1) VALUES (6, '(0.0 0.0');
                                                  ^
INSERT INTO POLYGON_TBL(ID,f1) VALUES (7, '(0,1,2)');
ERROR:  invalid input syntax for type polygon: "(0,1,2)"
LINE 1: INSERT INTO POLYGON_TBL(ID,f1) VALUES (7, '(0,1,2)');
                                                  ^
INSERT INTO POLYGON_TBL(ID,f1) VALUES (8, '(0,1,2,3');
ERROR:  invalid input syntax for type polygon: "(0,1,2,3"
LINE 1: INSERT INTO POLYGON_TBL(ID,f1) VALUES (8, '(0,1,2,3');
                                                  ^
INSERT INTO POLYGON_TBL(ID,f1) VALUES (9, 'asdf');
ERROR:  invalid input syntax for type polygon: "asdf"
LINE 1: INSERT INTO POLYGON_TBL(ID,f1) VALUES (9, 'asdf');
                                                  ^
SELECT '' AS four, * FROM POLYGON_TBL ORDER BY ID;
 four | id |         f1          
------+----+---------------------
      |  1 | ((2,0),(2,4),(0,0))
      |  2 | ((3,1),(3,3),(1,0))
      |  3 | ((0,0))
      |  4 | ((0,1),(0,1))
(4 rows)

-- overlap
SELECT '' AS three, p.*
   FROM POLYGON_TBL p
   WHERE p.f1 && '(3.0,1.0),(3.0,3.0),(1.0,0.0)' ORDER BY ID;
 three | id |         f1          
-------+----+---------------------
       |  1 | ((2,0),(2,4),(0,0))
       |  2 | ((3,1),(3,3),(1,0))
(2 rows)

-- left overlap
SELECT '' AS four, p.*
   FROM POLYGON_TBL p
   WHERE p.f1 &< '(3.0,1.0),(3.0,3.0),(1.0,0.0)' ORDER BY ID;
 four | id |         f1          
------+----+---------------------
      |  1 | ((2,0),(2,4),(0,0))
      |  2 | ((3,1),(3,3),(1,0))
      |  3 | ((0,0))
      |  4 | ((0,1),(0,1))
(4 rows)

-- right overlap
SELECT '' AS two, p.*
   FROM POLYGON_TBL p
   WHERE p.f1 &> '(3.0,1.0),(3.0,3.0),(1.0,0.0)' ORDER BY ID;
 two | id |         f1          
-----+----+---------------------
     |  2 | ((3,1),(3,3),(1,0))
(1 row)

-- left of
SELECT '' AS one, p.*
   FROM POLYGON_TBL p
   WHERE p.f1 << '(3.0,1.0),(3.0,3.0),(1.0,0.0)' ORDER BY ID;
 one | id |      f1       
-----+----+---------------
     |  3 | ((0,0))
     |  4 | ((0,1),(0,1))
(2 rows)

-- right of
SELECT '' AS zero, p.*
   FROM POLYGON_TBL p
   WHERE p.f1 >> '(3.0,1.0),(3.0,3.0),(1.0,0.0)' ORDER BY ID;
 zero | id | f1 
------+----+----
(0 rows)

-- contained
SELECT '' AS one, p.*
   FROM POLYGON_TBL p
   WHERE p.f1 <@ polygon '(3.0,1.0),(3.0,3.0),(1.0,0.0)' ORDER BY ID;
 one | id |         f1          
-----+----+---------------------
     |  2 | ((3,1),(3,3),(1,0))
(1 row)

-- same
SELECT '' AS one, p.*
   FROM POLYGON_TBL p
   WHERE p.f1 ~= polygon '(3.0,1.0),(3.0,3.0),(1.0,0.0)' ORDER BY ID;
 one | id |         f1          
-----+----+---------------------
     |  2 | ((3,1),(3,3),(1,0))
(1 row)

-- contains
SELECT '' AS one, p.*
   FROM POLYGON_TBL p
   WHERE p.f1 @> polygon '(3.0,1.0),(3.0,3.0),(1.0,0.0)' ORDER BY ID;
 one | id |         f1          
-----+----+---------------------
     |  2 | ((3,1),(3,3),(1,0))
(1 row)

--
-- polygon logic
--
-- left of
SELECT polygon '(2.0,0.0),(2.0,4.0),(0.0,0.0)' << polygon '(3.0,1.0),(3.0,3.0),(1.0,0.0)' AS false;
 false 
-------
 f
(1 row)

-- left overlap
SELECT polygon '(2.0,0.0),(2.0,4.0),(0.0,0.0)' << polygon '(3.0,1.0),(3.0,3.0),(1.0,0.0)' AS true;
 true 
------
 f
(1 row)

-- right overlap
SELECT polygon '(2.0,0.0),(2.0,4.0),(0.0,0.0)' &> polygon '(3.0,1.0),(3.0,3.0),(1.0,0.0)' AS false;
 false 
-------
 f
(1 row)

-- right of
SELECT polygon '(2.0,0.0),(2.0,4.0),(0.0,0.0)' >> polygon '(3.0,1.0),(3.0,3.0),(1.0,0.0)' AS false;
 false 
-------
 f
(1 row)

-- contained in
SELECT polygon '(2.0,0.0),(2.0,4.0),(0.0,0.0)' <@ polygon '(3.0,1.0),(3.0,3.0),(1.0,0.0)' AS false;
 false 
-------
 f
(1 row)

-- contains
SELECT polygon '(2.0,0.0),(2.0,4.0),(0.0,0.0)' @> polygon '(3.0,1.0),(3.0,3.0),(1.0,0.0)' AS false;
 false 
-------
 f
(1 row)

SELECT '((0,4),(6,4),(1,2),(6,0),(0,0))'::polygon @> '((2,1),(2,3),(3,3),(3,1))'::polygon AS "false";
 false 
-------
 f
(1 row)

SELECT '((0,4),(6,4),(3,2),(6,0),(0,0))'::polygon @> '((2,1),(2,3),(3,3),(3,1))'::polygon AS "true";
 true 
------
 t
(1 row)

SELECT '((1,1),(1,4),(5,4),(5,3),(2,3),(2,2),(5,2),(5,1))'::polygon @> '((3,2),(3,3),(4,3),(4,2))'::polygon AS "false";
 false 
-------
 f
(1 row)

SELECT '((0,0),(0,3),(3,3),(3,0))'::polygon @> '((2,1),(2,2),(3,2),(3,1))'::polygon AS "true";
 true 
------
 t
(1 row)

-- same
SELECT polygon '(2.0,0.0),(2.0,4.0),(0.0,0.0)' ~= polygon '(3.0,1.0),(3.0,3.0),(1.0,0.0)' AS false;
 false 
-------
 f
(1 row)

-- overlap
SELECT polygon '(2.0,0.0),(2.0,4.0),(0.0,0.0)' && polygon '(3.0,1.0),(3.0,3.0),(1.0,0.0)' AS true;
 true 
------
 t
(1 row)

SELECT '((0,4),(6,4),(1,2),(6,0),(0,0))'::polygon && '((2,1),(2,3),(3,3),(3,1))'::polygon AS "true";
 true 
------
 t
(1 row)

SELECT '((1,4),(1,1),(4,1),(4,2),(2,2),(2,4),(1,4))'::polygon && '((3,3),(4,3),(4,4),(3,4),(3,3))'::polygon AS "false";
 false 
-------
 f
(1 row)

SELECT '((200,800),(800,800),(800,200),(200,200))' &&  '(1000,1000,0,0)'::polygon AS "true";
 true 
------
 t
(1 row)

-- distance from a point
SELECT	'(0,0)'::point <-> '((0,0),(1,2),(2,1))'::polygon as on_corner,
	'(1,1)'::point <-> '((0,0),(2,2),(1,3))'::polygon as on_segment,
	'(2,2)'::point <-> '((0,0),(1,4),(3,1))'::polygon as inside,
	'(3,3)'::point <-> '((0,2),(2,0),(2,2))'::polygon as near_corner,
	'(4,4)'::point <-> '((0,0),(0,3),(4,0))'::polygon as near_segment;
 on_corner | on_segment | inside |   near_corner   | near_segment 
-----------+------------+--------+-----------------+--------------
         0 |          0 |      0 | 1.4142135623731 |          3.2
(1 row)

